### PR TITLE
WIP: Upgrade to composer 2.0

### PR DESCRIPTION
**Pull request by GromNaN**
This is a work in progress.

There is a massive amount of changes in the core classes of Composer that needs to be applied to Satis code. https://github.com/composer/composer/compare/1.10...master

#594

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -36,7 +37,6 @@
     },
     "require-dev": {
         "ext-json": "*",
-        "friendsofphp/php-cs-fixer": "^2.15",
```
Comment (alcohol) [inline composer.json:39]: Not sure I understand the reason for removing php-cs-fixer ? We use it for most of our linting and automated lint fixes.

Comment (alcohol) [reply composer.json:39]: I guess it conflicts with the composer packages?

Comment (GromNaN) [reply composer.json:39]: Yes, php-cs-fixer requires `"composer/semver": "^1.4"`.
https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/2.16/composer.json#L20

File: tests/Builder/PackagesBuilderDumpTest.php [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 use Composer\Json\JsonFile;
 use Composer\Package\Package;
+use Composer\Util\Loop;
```
Comment (stof) [inline tests/Builder/PackagesBuilderDumpTest.php:18]: why this change ?

**Comment by stof** [comment]
For now, this upgrade the package to use Composer 2 internally. That's not the same than outputting the new metadata format, right ?

**Comment by stof** [comment]
btw, for now, we should generate both the old and new format, so that the generated repository stays compatible with composer 1.

File: tests/Builder/PackagesBuilderDumpTest.php (hunk 2 above)
Comment (GromNaN) [reply tests/Builder/PackagesBuilderDumpTest.php:18]: An error.

**Comment by GromNaN** [comment]
For now, I'm just fighting to update to `composer/composer: 2.0@dev`. But I'm getting lost by the new async mecanism of the `HttpDownloader`.

**Comment by stof** [comment]
well, I think this PR should be considered separate from #594. We should not need to use composer 2 internally to output the new format (note that Packagist uses Composer 1 internally too)

File: tests/Builder/PackagesBuilderDumpTest.php (hunk 2 above)
Comment (andrerom) [reply tests/Builder/PackagesBuilderDumpTest.php:18]:
```suggestion
```

**Comment by andrerom** [comment]
Now that 2.0 is stable, once #619 is in place it might be feasible to finish this one right?

File: composer.json (hunk 1 above)
Comment (owenvoke) [reply composer.json:39]: It looks like `php-cs-fixer` can be re-added as it supports `composer/semver 2.x` since `v2.15.9` 👍🏻

**Comment by owenvoke** [comment]
Was this accidentally closed when the `master` branch was deleted (due to the base branch being `master`), or is development stopping on this? 🙂 If I can help at all with testing, let me know. 👍🏻

**Comment by alcohol** [comment]
I think it was closed because the `master` branch was removed, yes.

**Comment by ihor-sviziev** [comment]
@GromNaN, could you re-create a PR with the same changes to the `main` branch? Hope it will speed up the review and testing phase

**Comment by GromNaN** [comment]
I'm working on it.

**Comment by GromNaN** [comment]
New PR: https://github.com/composer/satis/pull/645
